### PR TITLE
bpo-36842: Implement PEP 578

### PR DESCRIPTION
**Pull request by zooba**
https://bugs.python.org/issue36842

### CODE REVIEW

File: Doc/library/urllib.request.rst
Comment (zooba) [inline Doc/library/urllib.request.rst]:
```suggestion
   ``urllib.request`` with arguments ``fullurl``, ``data``, ``headers``,
```

**Comment by zooba** [comment]
The PEP is not approved yet, of course, so the PR isn't finished. But I want it to at least run through CI now to catch anything that's going on.

**Comment by brettcannon** [comment]
Removing the import team from reviewing as this is still a WIP. We can be added back when it's ready to go.

**Comment by zooba** [comment]
Just waiting on a positive review - if you can contribute even part of a review, it would be much appreciated.

File: Include/internal/pycore_pystate.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_pystate.h:241]: This implies that audit hooks may be operating on Python objects outside an interpreter.  Is this intentional?  How will that work once we stop sharing the GIL?

File: Include/sysmodule.h
Comment (ericsnowcurrently) [inline Include/sysmodule.h]: I believe the convention now is to put this in `Include/cpython/sysmodule.h`.  Likewise for the other header additions.

File: Lib/importlib/_bootstrap_external.py
Comment (ericsnowcurrently) [inline Lib/importlib/_bootstrap_external.py:966]: This doesn't help with custom loaders that are impacted but don't inherit from those two loaders.  Does that matter?  If so, it may be worth adding an explicit attr/method for loaders to provide.

File: Objects/typeobject.c
Comment (ericsnowcurrently) [inline Objects/typeobject.c:396]: Isn't this for `type.__setattr__()`?

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [inline Python/pylifecycle.c:1254]: good comment

File: Python/pystate.c
Comment (ericsnowcurrently) [inline Python/pystate.c:48]: Are there plans to address this?

File: Doc/c-api/sys.rst
Comment (ericsnowcurrently) [inline Doc/c-api/sys.rst]: s/interpreter/runtime/ ?

File: Doc/c-api/sys.rst
Comment (ericsnowcurrently) [inline Doc/c-api/sys.rst]: It would help to have an explanation for each of the parameters sort of like you did for `sys.audit()`.

File: Doc/library/sys.rst
Comment (ericsnowcurrently) [inline Doc/library/sys.rst:71]: FWIW, I find the description for `args` to be rather unclear.

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c]: There isn't actually a `sys._clearaudithooks()` is there?

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c:266]: Based on what I've noticed elsewhere, this is a relatively sensitive function.  It may make sense to at least leave a comment to that effect here.

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c]: I'd expect something like `PyMem_RawFree()` here.

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c:297]: Would it make sense to cache the events until initialization has gotten far enough?

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c]: Shouldn't this be `PyMem_RawMalloc()`?

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c]: This appears to be entirely distinct from `sys.addaudithook()`.  Why the distinction?

The split seems to be on "global" vs. interpreter.  Which events can happen outside an interpreter?  I'd expect them to be few, mostly relating to initialization.  For everything else, why not support adding hooks to all interpreters?  Then the audit hook machinery could be driven from the interpreters, with the pre-init events handled with one-off code.

File: Python/sysmodule.c
Comment (ericsnowcurrently) [inline Python/sysmodule.c:177]: This is where we may get into trouble with objects crossing the interpreter boundary, right?

**Review by ericsnowcurrently — COMMENTED** [review]
I don't have any strong objections, but I do have a number of questions which I've mostly presented in my review comments.  The most significant relate to maintaining isolation between interpreters.

Also, it looks like there is a lot of new code all over the place.  What is the impact on performance?

File: Include/internal/pycore_pystate.h
Comment (tiran) [reply Include/internal/pycore_pystate.h:241]: The C audit hooks are defined globally but they are always called in the context of an interpreter and with a GIL held. The global hooks are shared between all sub-interpreters. Even the hooks for the main interpreter are only called inside the main interpreter.

https://www.python.org/dev/peps/pep-0578/#audit-hook may need additional clarification how the global hooks are called.

File: Python/sysmodule.c
Comment (tiran) [reply Python/sysmodule.c:297]: If you want to trace all audit events, then you have to define a global hook in C. The Python hooks are limited. This case may also need additional clarification in the PEP document.

File: Python/sysmodule.c
Comment (tiran) [reply Python/sysmodule.c]: First there are some events that can occur outside of an interpreter, e.g. initialization or creating of a new sub-interpreter. There are also a bunch of events that will occur before a "user-space" application gets a chance to register a hook. You have to load some modules and extension classes before you get a chance to do something.

For serious business, you have to implement and register a global hook before the interpreter is initialized. The ``sys.addaudithook()`` API is of limited use for security transparency.

File: Python/sysmodule.c
Comment (tiran) [reply Python/sysmodule.c:177]: I don't think so. The hooks will be called in the same thread as an interpreter with the GIL held. ``e->userData`` is not designed to hold or store any ``PyObject*`` data.

(This should be documented, too.)

File: Objects/typeobject.c
Comment (tiran) [reply Objects/typeobject.c:396]: Why is this hook ``object.__setatt__`` and not ``type.__stattr``?

File: Doc/library/os.rst
Comment (tiran) [inline Doc/library/os.rst:919]: Is this ``open`` or ``os.open``?

File: Include/sysmodule.h
Comment (tiran) [reply Include/sysmodule.h]: Is ``cpython/sysmodule.h`` designed to be used by 3rd party code? The idea is that 3rd parties are able to implement a custom Python binary that uses the hooks to audit the interpreter.

File: Lib/test/libregrtest/main.py
Comment (tiran) [inline Lib/test/libregrtest/main.py]: I think this should be guarded with ``if hasattr(sys, 'addaudithook')`` like in ``test_audit.py``

File: Lib/test/test_audit.py
Comment (tiran) [inline Lib/test/test_audit.py]: Could you please run black on the new tests?

File: Lib/test/test_embed.py
Comment (tiran) [inline Lib/test/test_embed.py:930]: guard with ``hasattr(sys, 'addaudithook')``?

File: Python/sysmodule.c
Comment (tiran) [inline Python/sysmodule.c:150]: How does the rest of the function work without a valid interpreter state? The function is operating on Python objects and at least calls ``PyTyple_New(0)``.

This function should verify that it's called with a valid interpreter state and the GIL held for the current interpreter. Otherwise it's calling for trouble and memory corruption.

**Review by tiran — DISMISSED** [review]

File: Lib/importlib/_bootstrap_external.py
Comment (zooba) [reply Lib/importlib/_bootstrap_external.py:966]: I don't want it to be trivially easy to disable the use of `open_code` here. Custom loaders need to opt-in to using it.

File: Objects/typeobject.c
Comment (zooba) [reply Objects/typeobject.c:396]: For consistency with the commented-out check in `object.c`, which raises `object.__setattr__` for all monkeypatching events (of course, since that one is disabled right now, perhaps this one could be named differently? Though it is still an object setattr, since the type object doesn't define its own one and just has special handling for some descriptors)

File: Python/pystate.c
Comment (zooba) [reply Python/pystate.c:48]: I hope so :)

I wasn't planning on restructuring runtime state in this PR.

File: Doc/c-api/sys.rst
Comment (zooba) [reply Doc/c-api/sys.rst]: Actually, this doc is just totally out of date...

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c]: No. I'll change it to use the C API name instead.

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c]: I forget why I had to use malloc/free, but I'm sure there was a reason. Maybe other things have changed by now so it's not necessary...

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c:297]: You don't receive any events until a hook is added, and you can't add a Python hook until you've initialized.

It might be more logical to check whether we're initialized before calling PySys_Audit in places where it matters, but it's much easier to do it here.

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c]: It's separate because a C hook (`void*`) behaves differently from a Python hook (`PyObject*`). We have to do both.

As it stands, we have prior art here. Setting a trace func or a profile func only applies to the current thread, and you have to manually add them to any new threads. For pure-Python hooks, doing it per-interpreter is a similar constraint, and similarly useful. It needs to be documented better here (I don't think it needs to be more clearly documented in the PEP - that's not where documentation lives for CPython).

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c:177]: Correct. The C function is not tied to a particular interpreter and will only be called in the context of the one where the arguments come from.

I'll improve the docs.

File: Lib/test/libregrtest/main.py
Comment (zooba) [reply Lib/test/libregrtest/main.py]: Yeah, probably. Just testing stuff right now to see whether it's worth having this on permanently

File: Doc/library/os.rst
Comment (zooba) [reply Doc/library/os.rst:919]: `open` is raised from a lot of places, regardless of how it's called. Distinguishing `os.open` from any other way a file is accessed just makes things harder for hooks.

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c:150]: It used to do that, then when the condition below was rewritten to deal with changes to pystate I updated it incorrectly (this is why I want to get this merged - there are so many changes that cause merge conflicts going on right now...)

File: Lib/test/test_embed.py
Comment (zooba) [reply Lib/test/test_embed.py:930]: We've failed to build test_embed if the hooks aren't available, so the whole test is broken. This file probably needs an "if not CPython: skip" condition.

Comment (zooba) [reply Lib/test/test_embed.py:930]: Which I'm not going to add right now, since we've clearly decided not to have one thus far :)

File: Include/internal/pycore_pystate.h
Comment (zooba) [reply Include/internal/pycore_pystate.h:241]: I added the clarifications to the docs, rather than the PEP, since that's where documentation lives now

File: Lib/test/libregrtest/main.py
Comment (zooba) [reply Lib/test/libregrtest/main.py]: Given the amount of stuff this is flushing out (I should have done it a while ago...), I'm going to leave it in. Hopefully it doesn't hurt the speed of the test runs too much...

File: Lib/test/libregrtest/main.py
Comment (vstinner) [inline Lib/test/libregrtest/main.py]: please don't add code at top-level. You can add code run in all tests in libregrtest/setup.py function.

File: Lib/test/libregrtest/save_env.py
Comment (vstinner) [inline Lib/test/libregrtest/save_env.py]: Why do you need this change? Could you maybe put in a separated PR if it's not directly related to PEP 578?

File: Modules/_ctypes/_ctypes.c
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c:3464]: This code is specific to Windows, no? #ifdef MS_WIN32 is needed somewhere.

File: Objects/fileobject.c
Comment (vstinner) [inline Objects/fileobject.c:38]: Can't you use io.open here? If there is a reason for using _io directly, maybe add a comment to explain why.

File: Python/sysmodule.c
Comment (zooba) [inline Python/sysmodule.c:199]: This is a big change that deserves review - we now disallow tracing inside hooks.

File: Python/fileutils.c
Comment (vstinner) [inline Python/fileutils.c]: I'm not sure that the GIL is held to call this function.

File: Doc/c-api/file.rst
Comment (vstinner) [inline Doc/c-api/file.rst]: Would it be possible to use a function pointer rather than void* here? void* may cause troubles with Control Flow Integrity (CFI).

File: Doc/howto/instrumentation.rst
Comment (vstinner) [inline Doc/howto/instrumentation.rst:338]: Does C string mean "char*"? What's the encoding of the event name? UTF-8 I guess? Or ASCII?

File: Python/ceval.c
Comment (vstinner) [inline Python/ceval.c]: Since Python 3.6, we use C99 and so it's fine to add code before variable declarations.

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: Can you only clear the exception if audit fails?

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c:255]: Can you please add: assert(PyErr_Occurred());

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: Maybe you could use METH_FASTCALL to avoid the temporary tuple to pass positional arguments.

File: Objects/fileobject.c
Comment (tiran) [reply Objects/fileobject.c:38]: It must be ``_io`` because import of ``io`` module may cause a cascade of additional auditing events.

File: Doc/howto/instrumentation.rst
Comment (tiran) [reply Doc/howto/instrumentation.rst:338]: This is a USDT probe point, not actually a C function. We just use ``c:function`` to encode USDT. It's a NULL terminated string array that contains a user space string.

File: Lib/test/libregrtest/save_env.py
Comment (zooba) [reply Lib/test/libregrtest/save_env.py]: It was directly related to figuring out which thread was leaking. I don't need it anymore, so either I can revert it or leave it - I'm not filing a new bug for it.

File: Objects/fileobject.c
Comment (zooba) [reply Objects/fileobject.c:38]: Yep, we may need this function to open `io.py` in order to import it.

File: Python/fileutils.c
Comment (zooba) [reply Python/fileutils.c]: I've moved the auditing for `_Py_wfopen` and `_Py_fopen` to the call sites, since most of them do actually hold the GIL.

File: Python/sysmodule.c
Comment (zooba) [reply Python/sysmodule.c]: Do you have any docs on how to implement METH_FASTCALL that will help me here?

File: Python/fileutils.c
Comment (vstinner) [reply Python/fileutils.c]: I see many calls in Modules/getpath.c and PC/getpathp.c. In these files, there is no interpreter yet nor GIL, no? I only see one call in pymain_run_file() which should hold the GIL and have an intererpreter running.

Comment (vstinner) [reply Python/fileutils.c]: Well, PySys_Audit() should detect if there is an interpreter or not. It is just a remark :-)

**Comment by vstinner** [comment]
Patch to use METH_FASTCALL:

```
diff --git a/Python/sysmodule.c b/Python/sysmodule.c
index f987bb2016..fd1408c748 100644
--- a/Python/sysmodule.c
+++ b/Python/sysmodule.c
@@ -22,6 +22,7 @@ Data members:
 #include "pycore_pymem.h"
 #include "pycore_pathconfig.h"
 #include "pycore_pystate.h"
+#include "pycore_tupleobject.h"
 #include "pythread.h"
 #include "pydtrace.h"
 
@@ -360,20 +361,14 @@ PyDoc_STRVAR(audit_doc,
 Passes the event to any audit hooks that are attached.");
 
 static PyObject *
-sys_audit(PyObject *self, PyObject *args)
+sys_audit(PyObject *self, PyObject *const *args, Py_ssize_t argc)
 {
-    if (!PyTuple_Check(args)) {
-        PyErr_Format(PyExc_TypeError, "expected tuple, not %.200s", Py_TYPE(args)->tp_name);
-        return NULL;
-    }
-
-    Py_ssize_t argc = PyTuple_GET_SIZE(args);
     if (argc == 0) {
         PyErr_SetString(PyExc_TypeError, "audit() missing 1 required positional argument: 'event'");
         return NULL;
     }
 
-    PyObject *auditEvent = PyTuple_GET_ITEM(args, 0);
+    PyObject *auditEvent = args[0];
     if (!auditEvent) {
         PyErr_SetString(PyExc_TypeError, "expected str for argument 'event'");
         return NULL;
@@ -388,16 +383,10 @@ sys_audit(PyObject *self, PyObject *args)
         return NULL;
     }
 
-    PyObject *auditArgs = PyTuple_New(argc - 1);
+    PyObject *auditArgs = _PyTuple_FromArray(args + 1, argc - 1);
     if (!auditArgs) {
         return NULL;
     }
-    for (int i = 0; i < argc - 1; ++i) {
-        PyObject *o = PyTuple_GET_ITEM(args, i + 1);
-        Py_INCREF(o);
-        PyTuple_SET_ITEM(auditArgs, i, o);
-    }
-
     int res = PySys_Audit(event, "O", auditArgs);
     Py_DECREF(auditArgs);
 
@@ -1924,7 +1913,7 @@ sys_getandroidapilevel_impl(PyObject *module)
 static PyMethodDef sys_methods[] = {
     /* Might as well keep this in alphabetic order */
     SYS_ADDAUDITHOOK_METHODDEF
-    {"audit",           sys_audit, METH_VARARGS, audit_doc },
+    {"audit",           (PyCFunction)(void(*)(void))sys_audit, METH_FASTCALL, audit_doc },
     {"breakpointhook",  (PyCFunction)(void(*)(void))sys_breakpointhook,
      METH_FASTCALL | METH_KEYWORDS, breakpointhook_doc},
     SYS_CALLSTATS_METHODDEF
```

By the way, I might be useful to extract the code from PySys_Audit() to decide if audit is used or not: create a subfunction, and call it from sys_audit() to do nothing if audit is not used (common case).

Calling PyUnicode_AsUTF8() and _PyTuple_FromArray() is not free :-) (I know that they are fast, but many function calls take less than 100 ns overall.)

File: Lib/test/libregrtest/save_env.py
Comment (zooba) [reply Lib/test/libregrtest/save_env.py]: To be more explicit, `WeakSet.__repr__` does not show the contents, so when a dangling thread remains this didn't show anything useful.

File: Python/fileutils.c
Comment (zooba) [reply Python/fileutils.c]: You're right, if DTrace is enabled then we'll get past the other checks. I'll fix that up.

Hopefully in the future we can reenable primitive types outside of a specific interpreter and then won't have to skip early events (some of us are already thinking about this, but it's not going to happen right now).

Comment (zooba) [reply Python/fileutils.c]: And I'll put the audit call back into these functions, since it'll have no effect when the GIL is not held (until the point where it can have an effect)

**Comment by zooba** [comment]
@tiran I believe all feedback has been addressed, and all tests pass (custom buildbot run going now). Anything else?

File: Lib/test/libregrtest/save_env.py
Comment (zooba) [reply Lib/test/libregrtest/save_env.py]: Reverted this and filed bpo-36949

**Comment by zooba** [comment]
I have made the requested changes; please review again

**Comment by zooba** [comment]
@tiran I know you're busy with your PEP, but this is waiting on you.

I'll give it two more merge conflicts from other commits before I consider your feedback addressed and merge :)

**Comment by zooba** [comment]
That's the second conflict, so I'm dismissing @tiran's review and we can deal with any further issues post-commit.

**Pull request merged**
